### DISCUSSION TITLE
About max_pool

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#2641] [Q&A] [answered]
Hi Matt, I have some questions about the code below. 

``` python
def forward(self, data):
    data.x = F.elu(self.conv1(data.x, data.edge_index, data.edge_attr))
    weight = normalized_cut_2d(data.edge_index, data.pos)
    cluster = graclus(data.edge_index, weight, data.x.size(0))
    data.edge_attr = None
    data = max_pool(cluster, data, transform=transform)

    data.x = F.elu(self.conv2(data.x, data.edge_index, data.edge_attr))
    weight = normalized_cut_2d(data.edge_index, data.pos)
    cluster = graclus(data.edge_index, weight, data.x.size(0))
    x, batch = max_pool_x(cluster, data.x, data.batch)

    x = global_mean_pool(x, batch)
    x = F.elu(self.fc1(x))
    x = F.dropout(x, training=self.training)
    return F.log_softmax(self.fc2(x), dim=1)
```

1. What is the difference between `max_pool` and `max_pool_x`?

2. Why should we reset `data.edge_attr` using `data.edge_attr=None` before `max_pool`?

**Comment by rusty1s** [comment, marked as answer]
1. `max_pool` will pool a given `data` object, i.e., it will take care of coarsening the graph and pool node features, while `max_pool_x` will only take care of pooling node features (e.g., in case `edge_index` is no longer needed at this point of time).
2. You don't have to, but here we compute new edge features (the spatial relations between nodes) given the new graph representation.
